[scroll-animations] refactor code related to `animation-timeline` and `timeline-scope` out of `AnimationTimelinesController`

graouts opened the pull request:
#### 801565e14227f7d3a2fc7a7fc6d3fe3ec3d108ba
<pre>
[scroll-animations] refactor code related to `animation-timeline` and `timeline-scope` out of `AnimationTimelinesController`
<a href="https://bugs.webkit.org/show_bug.cgi?id=287767">https://bugs.webkit.org/show_bug.cgi?id=287767</a>

Reviewed by Anne van Kesteren.

We&apos;ve been adding a fair bit of code to `AnimationTimelinesController` since starting work on Scroll-driven
Animations to support the `animation-timeline` and `timeline-scope` properties. Since there is almost no
overlap between the pre-existing functionality of `AnimationTimelinesController` and the added code, it
makes sense to split off this new code in a dedicated class: `ScrollOriginatedTimelinesController`.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::isPendingTimelineAttachment const):
(WebCore::originatingElement): Deleted.
(WebCore::originatingStyleableIncludingTimelineScope): Deleted.
(WebCore::originatingElementExcludingTimelineScope): Deleted.
(WebCore::AnimationTimelinesController::relatedTimelineScopeElements): Deleted.
(WebCore::AnimationTimelinesController::inactiveNamedTimeline): Deleted.
(WebCore::containsElement): Deleted.
(WebCore::AnimationTimelinesController::determineTreeOrder): Deleted.
(WebCore::AnimationTimelinesController::determineTimelineForElement): Deleted.
(WebCore::AnimationTimelinesController::timelinesForName): Deleted.
(WebCore::AnimationTimelinesController::updateTimelineForTimelineScope): Deleted.
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline): Deleted.
(WebCore::AnimationTimelinesController::updateCSSAnimationsAssociatedWithNamedTimeline): Deleted.
(WebCore::AnimationTimelinesController::removePendingOperationsForCSSAnimation): Deleted.
(WebCore::AnimationTimelinesController::documentDidResolveStyle): Deleted.
(WebCore::AnimationTimelinesController::registerNamedViewTimeline): Deleted.
(WebCore::AnimationTimelinesController::unregisterNamedTimeline): Deleted.
(WebCore::AnimationTimelinesController::setTimelineForName): Deleted.
(WebCore::updateTimelinesForTimelineScope): Deleted.
(WebCore::AnimationTimelinesController::updateNamedTimelineMapForTimelineScope): Deleted.
(WebCore::AnimationTimelinesController::unregisterNamedTimelinesAssociatedWithElement): Deleted.
(WebCore::AnimationTimelinesController::styleableWasRemoved): Deleted.
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncStyleOriginatedTimeline):
* Source/WebCore/animation/StyleOriginatedTimelinesController.cpp: Copied from Source/WebCore/animation/AnimationTimelinesController.cpp.
(WebCore::originatingElement):
(WebCore::originatingStyleableIncludingTimelineScope):
(WebCore::originatingElementExcludingTimelineScope):
(WebCore::StyleOriginatedTimelinesController::relatedTimelineScopeElements):
(WebCore::StyleOriginatedTimelinesController::inactiveNamedTimeline):
(WebCore::containsElement):
(WebCore::StyleOriginatedTimelinesController::determineTreeOrder):
(WebCore::StyleOriginatedTimelinesController::determineTimelineForElement):
(WebCore::StyleOriginatedTimelinesController::timelinesForName):
(WebCore::StyleOriginatedTimelinesController::updateTimelineForTimelineScope):
(WebCore::StyleOriginatedTimelinesController::registerNamedScrollTimeline):
(WebCore::StyleOriginatedTimelinesController::updateCSSAnimationsAssociatedWithNamedTimeline):
(WebCore::StyleOriginatedTimelinesController::removePendingOperationsForCSSAnimation):
(WebCore::StyleOriginatedTimelinesController::documentDidResolveStyle):
(WebCore::StyleOriginatedTimelinesController::registerNamedViewTimeline):
(WebCore::StyleOriginatedTimelinesController::unregisterNamedTimeline):
(WebCore::StyleOriginatedTimelinesController::setTimelineForName):
(WebCore::updateTimelinesForTimelineScope):
(WebCore::StyleOriginatedTimelinesController::updateNamedTimelineMapForTimelineScope):
(WebCore::StyleOriginatedTimelinesController::isPendingTimelineAttachment const):
(WebCore::StyleOriginatedTimelinesController::unregisterNamedTimelinesAssociatedWithElement):
(WebCore::StyleOriginatedTimelinesController::styleableWasRemoved):
* Source/WebCore/animation/StyleOriginatedTimelinesController.h: Copied from Source/WebCore/animation/AnimationTimelinesController.h.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):
(WebCore::Document::ensureStyleOriginatedTimelinesController):
* Source/WebCore/dom/Document.h:
(WebCore::Document::styleOriginatedTimelinesController):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::elementWasRemoved const):
(WebCore::Styleable::cancelStyleOriginatedAnimations const):
(WebCore::Styleable::updateCSSScrollTimelines const):
(WebCore::Styleable::updateCSSViewTimelines const):

Canonical link: <a href="https://commits.webkit.org/290462@main">https://commits.webkit.org/290462@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d1fb7b9bf8e2ba633244a11c83643c967818eb28

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90136 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9665 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45051 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/95137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40912 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92188 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10052 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17973 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/95137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/26996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93137 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/7701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/81771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/95137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/40043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/37198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/96962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17324 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/96962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17581 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/77595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/96962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/22049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10558 "Failed to checkout and rebase branch from PR 40682") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14168 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17334 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22660 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17075 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20527 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18859 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/animation/StyleOriginatedTimelinesController.h:
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
```
Shouldn't this start at 2020? Prolly same for the other file.

graouts replied:
You're right, I'll fix it.

smfr commented on Source/WebCore/animation/StyleOriginatedTimelinesController.h:
```diff
@@ -0,0 +1,86 @@
+};
+
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(StyleOriginatedTimelinesController);
+class StyleOriginatedTimelinesController final : public CanMakeCheckedPtr<StyleOriginatedTimelinesController> {
```
This would be a great place to add a brief comment to explain what a "style-originated" timeline is. Does it include CSS Animation timelines?

graouts replied:
Good point! A style-originated timeline is a timeline that is assigned to a CSS Animation via the `animation-timeline` property and which may also involve the `scroll-timeline-name`, `view-timeline-name` and `timeline-scope` properties. I'll make sure to add this as a comment.

graouts committed:
[scroll-animations] refactor code related to `animation-timeline` and `timeline-scope` out of `AnimationTimelinesController`
https://bugs.webkit.org/show_bug.cgi?id=287767

Reviewed by Anne van Kesteren.

We've been adding a fair bit of code to `AnimationTimelinesController` since starting work on Scroll-driven
Animations to support the `animation-timeline` and `timeline-scope` properties. Since there is almost no
overlap between the pre-existing functionality of `AnimationTimelinesController` and the added code, it
makes sense to split off this new code in a dedicated class: `ScrollOriginatedTimelinesController`.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::isPendingTimelineAttachment const):
(WebCore::originatingElement): Deleted.
(WebCore::originatingStyleableIncludingTimelineScope): Deleted.
(WebCore::originatingElementExcludingTimelineScope): Deleted.
(WebCore::AnimationTimelinesController::relatedTimelineScopeElements): Deleted.
(WebCore::AnimationTimelinesController::inactiveNamedTimeline): Deleted.
(WebCore::containsElement): Deleted.
(WebCore::AnimationTimelinesController::determineTreeOrder): Deleted.
(WebCore::AnimationTimelinesController::determineTimelineForElement): Deleted.
(WebCore::AnimationTimelinesController::timelinesForName): Deleted.
(WebCore::AnimationTimelinesController::updateTimelineForTimelineScope): Deleted.
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline): Deleted.
(WebCore::AnimationTimelinesController::updateCSSAnimationsAssociatedWithNamedTimeline): Deleted.
(WebCore::AnimationTimelinesController::removePendingOperationsForCSSAnimation): Deleted.
(WebCore::AnimationTimelinesController::documentDidResolveStyle): Deleted.
(WebCore::AnimationTimelinesController::registerNamedViewTimeline): Deleted.
(WebCore::AnimationTimelinesController::unregisterNamedTimeline): Deleted.
(WebCore::AnimationTimelinesController::setTimelineForName): Deleted.
(WebCore::updateTimelinesForTimelineScope): Deleted.
(WebCore::AnimationTimelinesController::updateNamedTimelineMapForTimelineScope): Deleted.
(WebCore::AnimationTimelinesController::unregisterNamedTimelinesAssociatedWithElement): Deleted.
(WebCore::AnimationTimelinesController::styleableWasRemoved): Deleted.
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncStyleOriginatedTimeline):
* Source/WebCore/animation/StyleOriginatedTimelinesController.cpp: Copied from Source/WebCore/animation/AnimationTimelinesController.cpp.
(WebCore::originatingElement):
(WebCore::originatingStyleableIncludingTimelineScope):
(WebCore::originatingElementExcludingTimelineScope):
(WebCore::StyleOriginatedTimelinesController::relatedTimelineScopeElements):
(WebCore::StyleOriginatedTimelinesController::inactiveNamedTimeline):
(WebCore::containsElement):
(WebCore::StyleOriginatedTimelinesController::determineTreeOrder):
(WebCore::StyleOriginatedTimelinesController::determineTimelineForElement):
(WebCore::StyleOriginatedTimelinesController::timelinesForName):
(WebCore::StyleOriginatedTimelinesController::updateTimelineForTimelineScope):
(WebCore::StyleOriginatedTimelinesController::registerNamedScrollTimeline):
(WebCore::StyleOriginatedTimelinesController::updateCSSAnimationsAssociatedWithNamedTimeline):
(WebCore::StyleOriginatedTimelinesController::removePendingOperationsForCSSAnimation):
(WebCore::StyleOriginatedTimelinesController::documentDidResolveStyle):
(WebCore::StyleOriginatedTimelinesController::registerNamedViewTimeline):
(WebCore::StyleOriginatedTimelinesController::unregisterNamedTimeline):
(WebCore::StyleOriginatedTimelinesController::setTimelineForName):
(WebCore::updateTimelinesForTimelineScope):
(WebCore::StyleOriginatedTimelinesController::updateNamedTimelineMapForTimelineScope):
(WebCore::StyleOriginatedTimelinesController::isPendingTimelineAttachment const):
(WebCore::StyleOriginatedTimelinesController::unregisterNamedTimelinesAssociatedWithElement):
(WebCore::StyleOriginatedTimelinesController::styleableWasRemoved):
* Source/WebCore/animation/StyleOriginatedTimelinesController.h: Copied from Source/WebCore/animation/AnimationTimelinesController.h.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):
(WebCore::Document::ensureStyleOriginatedTimelinesController):
* Source/WebCore/dom/Document.h:
(WebCore::Document::styleOriginatedTimelinesController):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::elementWasRemoved const):
(WebCore::Styleable::cancelStyleOriginatedAnimations const):
(WebCore::Styleable::updateCSSScrollTimelines const):
(WebCore::Styleable::updateCSSViewTimelines const):

Canonical link: https://commits.webkit.org/290462@main
